File: Second Year/1st Semester/Metode Avansate De Programare/Laboratoare/LABORATOR_3/Laborator_3/src/repository/memory/InMemoryRepository.java
package repository.memory;

import domain.Entity;
import domain.validator.ValidationException;
import domain.validator.Validator;
import jdk.jshell.spi.ExecutionControl;
import repository.Repository;

import java.util.HashMap;
import java.util.Map;

public abstract class InMemoryRepository<ID, E extends Entity<ID>> implements Repository<E,ID> {
    private Validator<E> validator;
    private Map<ID,E> entities;

    public InMemoryRepository(Validator<E> validator) {
        this.validator = validator;
        entities=new HashMap<ID,E>();
    }

    @Override
    public E findOne(ID e) {
        if(e == null)
            throw new IllegalArgumentException("The id can not be null");
        else
        {
            E entity = null;
            for(E ent : entities.values())
            {
                if(ent.getId() == e)
                    entity = ent;
            }
            return entity;
        }
    }

    @Override
    public Iterable<E> findAll() {
        return entities.values();
    }

    @Override
    public E save(E entity) {
        if (entity == null) {
            throw new IllegalArgumentException("Entity cannot be null");
        }
        if (entities.containsKey(entity.getId())) {
            return entity;
        }
        validator.validate(entity);
        entities.put(entity.getId(), entity);
        return entity;
    }

    @Override
    public E delete(ID e) {
        if (e == null) {
            throw new IllegalArgumentException("Entity cannot be null");
        }
        return entities.remove(e);
    }

    @Override
    public E update(E entity) {
        if (entity == null) {
            throw new IllegalArgumentException("Entity cannot be null");
        }
        try{
            validator.validate(entity);
        }
        catch (ValidationException E)
        {
            System.out.println("The entity is not valid!");
        }
        return entities.put(entity.getId(), entity);
    }
}

File: Second Year/1st Semester/Metode Avansate De Programare/Laboratoare/LABORATOR_8/LABORATOR_8/Laborator_C_sharp/Repository/InMemoryRepository.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Laborator_C_sharp.Domain;

namespace Laborator_C_sharp.Repository
{
    class InMemoryRepository<ID,E> : Repository<ID,E> where E : Entity<ID>
    {
        protected IDictionary<ID, E> entities = new Dictionary<ID, E>();
        public E Delete(ID id)
        {
            if (!this.entities.ContainsKey(id))
                return null;
            E entity = this.entities[id];
            this.entities.Remove(id);
            return entity;
        }
        public IEnumerable<E> FindAll()
        {
            return this.entities.Values.ToList<E>();
        }
        public E FindOne(ID id)
        {
            if(this.entities.ContainsKey(id))
                return this.entities[id];
            return null;
        }
        public E Save(E entity)
        {
            if (entity == null)
                throw new ArgumentNullException("Entity must not be null");
            if (this.entities.ContainsKey(entity.Id))
                return entity;
            this.entities[entity.Id] = entity;
            return default;
        }
        public E Update(E entity)
        {
            if(this.entities.ContainsKey(entity.Id))
            {
                this.entities[entity.Id] = entity;
                return entity; 
            }
            return null;
        }
    }
}
